git/libgit2/transport/http_test.go

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	git2go "github.com/libgit2/git2go/v33"
+	. "github.com/onsi/gomega"
+
+	"github.com/fluxcd/pkg/git"
+	"github.com/fluxcd/pkg/git/libgit2/internal/test"
+	"github.com/fluxcd/pkg/gittestserver"
+)
+
+func TestHttpAction_CreateClientRequest(t *testing.T) {
+	authOpts := git.AuthOptions{
+		Username: "user",
+		Password: "pwd",
+	}
+	url := "https://final-target/abc"
+
+	tests := []struct {
+		name       string
+		assertFunc func(g *WithT, req *http.Request, client *http.Client)
+		action     git2go.SmartServiceAction
+		authOpts   git.AuthOptions
+		transport  *http.Transport
+		wantedErr  error
+	}{
+		{
+			name:   "Uploadpack: URL, method and headers are correctly set",
+			action: git2go.SmartServiceActionUploadpack,
+			transport: &http.Transport{
+				Proxy:              http.ProxyFromEnvironment,
+				ProxyConnectHeader: map[string][]string{},
+			},
+			assertFunc: func(g *WithT, req *http.Request, _ *http.Client) {
+				g.Expect(req.URL.String()).To(Equal("https://final-target/abc/git-upload-pack"))
+				g.Expect(req.Method).To(Equal("POST"))
+				g.Expect(req.Header).To(BeEquivalentTo(map[string][]string{
+					"User-Agent":   {"git/2.0 (flux-libgit2)"},
+					"Content-Type": {"application/x-git-upload-pack-request"},
```

Comment:
It may be better to create more constants for all these different types of content types, long strings, error-prone.